Recognize autofill as a user interaction

sysrqb opened the pull request:
#### fdc846a6dcfb6f978308992990890ad587c43795
<pre>
Recognize autofill as a user interaction
<a href="https://bugs.webkit.org/show_bug.cgi?id=290788">https://bugs.webkit.org/show_bug.cgi?id=290788</a>
<a href="https://rdar.apple.com/148268327">rdar://148268327</a>

Reviewed by Aditya Keerthi.

Currently autofill is not logged as a user interaction, despite it requiring
explicit user consent. This patch resolves that problem by logging a user
interaction on form autofill.

* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-obscured-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-obscured-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-viewable-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-viewable-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-user-interaction.html: Added.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::logUserInteraction):
(WebCore::HTMLInputElement::setAutofilled):
(WebCore::HTMLInputElement::setAutofilledAndViewable):
(WebCore::HTMLInputElement::setAutofilledAndObscured):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: <a href="https://commits.webkit.org/293263@main">https://commits.webkit.org/293263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4253e3d91b8a077e0582724219b057ed146758e8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98122 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17753 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7980 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103238 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48651 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100167 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26204 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74708 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31890 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13674 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88668 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55068 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13458 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6602 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48093 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83435 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6682 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105615 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25208 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18368 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83696 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84848 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83149 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/21009 "Build was cancelled. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Checked out pull request; Reviewed by Aditya Keerthi; Compiled WebKit (cancelled)") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27787 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5477 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18848 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15938 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25167 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30341 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24987 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1532,11 +1533,21 @@ URL HTMLInputElement::src() const
+{
+    RefPtr mainFrame = document().frame() && document().frame()->localMainFrame() ? document().frame()->localMainFrame() : nullptr;
+    if (RefPtr document = mainFrame ? mainFrame->document() : nullptr)
+        ResourceLoadObserver::shared().logUserInteractionWithReducedTimeResolution(*document);
```
I'd have written this with early returns, rather than using `nullptr` twice to drive the control flow. It would also remove the duplication of `document().frame()` on the first line.

sysrqb replied:
Thanks, made those changes.

pxlcoder reviewed (approved):
Any chance this is testable? Note that we can already call these autofill methods using `window.internals`.

Commit:
Recognize autofill as a user interaction
https://bugs.webkit.org/show_bug.cgi?id=290788
rdar://148268327

Reviewed by Aditya Keerthi.

Currently autofill is not logged as a user interaction, despite it requiring
explicit user consent. This patch resolves that problem by logging a user
interaction on form autofill.

* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-obscured-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-obscured-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-viewable-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-and-viewable-user-interaction.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-user-interaction-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/input-autofilled-user-interaction.html: Added.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::logUserInteraction):
(WebCore::HTMLInputElement::setAutofilled):
(WebCore::HTMLInputElement::setAutofilledAndViewable):
(WebCore::HTMLInputElement::setAutofilledAndObscured):
* Source/WebCore/html/HTMLInputElement.h:

Canonical link: https://commits.webkit.org/293263@main
